shopelectro/views/catalog.py

```diff
@@ -25,6 +26,62 @@ def get_products_count(request):
     return PRODUCTS_ON_PAGE_MOB if mobile_view else PRODUCTS_ON_PAGE_PC
 
 
+def get_view_type(request):
+    view_type = request.session.get('view_type', 'tile')
+    assert view_type in ['list', 'tile']
+    return view_type
+
+
+def get_catalog_context(request, category, raw_tags, page_number, per_page, sorting_index):
+    all_tags = newcontext.Tags(models.Tag.objects.all())
+
+    if raw_tags:
+        selected_tags = newcontext.tags.Checked404Tags(
+            newcontext.tags.ParsedTags(
+                tags=all_tags,
+                raw_tags=raw_tags,
+            )
+        )
+    else:
+        selected_tags = newcontext.Tags(models.Tag.objects.none())
+
+    products = newcontext.products.OrderedProducts(
+        products=newcontext.products.TaggedProducts(
+            products=newcontext.products.ProductsByCategory(
+                products=newcontext.products.ActiveProducts(
+                    newcontext.Products(
+                        models.Product.objects.all(),
+                    ),
+                ),
+                category=category,
+            ),
+            tags=selected_tags,
+        ),
+        sorting_index=sorting_index,
+    )
+
+    paginated_products = newcontext.products.PaginatedProducts(
+        products=products,
+        url=request.path,
+        page_number=page_number,
+        per_page=per_page,
+    )
+
+    images = newcontext.products.ProductImages(paginated_products, Image.objects.all())
+    brands = newcontext.products.ProductBrands(paginated_products, all_tags)
+    grouped_tags = newcontext.tags.GroupedTags(all_tags)
+
+    contexts = newcontext.Contexts(
+        all_tags, paginated_products, images, brands, grouped_tags,
+    )
+    optional_context = {
+        'skip_canonical': selected_tags.qs().exists(),
+        'total_products': products.qs().count(),
+        'selected_tags': selected_tags.qs(),
+    }
+    return contexts, optional_context
```

Comment:
plz provide comment why you are not chaining optional_context to contexts object

Reply:
It will be removed in a scope of next tasks

Reply:
@artemiy312 , so write it in comment plz =)

Reply:
I have already created a todo with an exhaustive description for it